[PyTorch] Use plain old function pointer for RecordFunctionCallback (reapply)

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49408 [PyTorch] Use plain old function pointer for RecordFunctionCallback (reapply)**

Nearly every non-test callsite doesn't need to capture any variables anyway, and this saves 48 bytes per callback.

Differential Revision: [D25563207](https://our.internmc.facebook.com/intern/diff/D25563207/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25563207/)!

swolchok committed:
[PyTorch] Use plain old function pointer for RecordFunctionCallback (reapply)

Nearly every non-test callsite doesn't need to capture any variables anyway, and this saves 48 bytes per callback.

Differential Revision: [D25563207](https://our.internmc.facebook.com/intern/diff/D25563207/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25563207/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use plain old function pointer for RecordFunctionCallback (reapply)"


Nearly every non-test callsite doesn't need to capture any variables anyway, and this saves 48 bytes per callback.

Differential Revision: [D25563207](https://our.internmc.facebook.com/intern/diff/D25563207/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25563207/)!

[ghstack-poisoned]

malfet commented on aten/src/ATen/record_function.cpp:
```diff
@@ -277,10 +277,12 @@ class CallbackManager {
     try {
       if (is_start) {
-        ctx = rfcb.start()(rf);
+        ctx = rfcb.start() ? rfcb.start()(rf) : nullptr;
```
How about something like?
```suggestion
        if (auto& start = rcbf.start()) ctx = start(rf) else ctx =  nullptr;
```

swolchok replied:
there's no reason to complicate the code like this; start() is an inlineable load of a single pointer

malfet commented on aten/src/ATen/record_function.cpp:
```diff
@@ -277,10 +277,12 @@ class CallbackManager {
       }
       else {
-        rfcb.end()(rf, ctx.get());
+        if (rfcb.end()) {
+          rfcb.end()(rf, ctx.get());
+        }
```
Nit
```suggestion
        if (auto& end = rfcb.end()) {
          end(rf, ctx.get());
        }
```

malfet commented on aten/src/ATen/record_function.h:
```diff
@@ -305,14 +305,16 @@ struct TORCH_API RecordFunction {
-        [](const RecordFunction&, ObserverContext*) {}):
-      start_(std::move(start)),
-      end_(std::move(end)) {
+      StartCallback start,
```
Why not pass by reference here?(or `keep std::move()` when member variable is initialized
```suggestion
      const StartCallback& start,
```

swolchok replied:
because function pointers should be passed by value, just like `void*` or `int`, for efficiency reasons. Similarly, there is no need to move them.

malfet commented on aten/src/ATen/record_function.h:
```diff
@@ -368,18 +370,18 @@ class TORCH_API RecordFunctionCallback {
   }
 
-  inline const std::function<std::unique_ptr<ObserverContext>(const RecordFunction&)>& start() const {
+  inline StartCallback start() const {
```
Why do we want a copy here?
```suggestion
  inline const StartCallback& start() const {
```

swolchok replied:
because function pointers are cheap to copy, like `int` or `void*`

malfet replied:
`std::function` is not just a function pointer, is it?

swolchok replied:
No, but this diff changes StartCallback to be a function pointer above.

malfet commented on binaries/record_function_benchmark.cc:
```diff
@@ -19,10 +19,10 @@ const float kLowSamplingProb = 0.0001;
     double sampling_prob = 1.0,
-    std::function<std::unique_ptr<at::ObserverContext>(const at::RecordFunction&)> fn =
-        [](const at::RecordFunction&) { return nullptr; }) {
+    at::RecordFunctionCallback::StartCallback fn =
+        [](const at::RecordFunction&) -> std::unique_ptr<at::ObserverContext> { return nullptr; }) {
```
Why do we need a non-trivial initializer here?
```suggestion
    at::RecordFunctionCallback::StartCallback fn = at::RecordFunctionCallback::StartCallback()
```

swolchok replied:
I just left it the way the existing code had it, don't think it's expensive either way because it's a constant expression (address of a function)

swolchok committed:
Update on "[PyTorch] Use plain old function pointer for RecordFunctionCallback (reapply)"


Nearly every non-test callsite doesn't need to capture any variables anyway, and this saves 48 bytes per callback.

Differential Revision: [D25563207](https://our.internmc.facebook.com/intern/diff/D25563207/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25563207/)!

[ghstack-poisoned]
